### ISSUE TITLE
HSL IO performance: add native awaitables for looping operations like writeAllAsync

### ISSUE THREAD

**Issue by fredemmott** [hhvm/hsl-experimental#137]
e.g. writeAllAsync, copying between FDs

These can be implemented in pure hack, however given limitations of epoll, we currently do async IO in background threads - switching threads every time we go through the loop has a fair bit of overhead. A native awaitable could run the whole lot in a worker thread, and only re-enter the hack code once the entire operation has failed or is finished.

This is not needed for completeness, just a performance optimization.
